File: android/app/src/main/java/com/noop/analytics/StressHeatmap.kt
package com.noop.analytics

/**
 * StressHeatmap — pure week×hour aggregation of the intraday autonomic-load timeline (R7).
 *
 * Kotlin twin of the Swift `StressHeatmap`; keep every cut point, ordering rule and tie-break
 * BYTE-IDENTICAL so iOS and Android resolve the same grid + pattern from the same inputs. No I/O, no
 * Android types — unit-testable on the JVM.
 *
 * Answers the pattern question the daily score and today's timeline can't: which hours of which days run
 * habitually loaded ("Wednesday afternoons run hot"). Derived entirely from data NOOP already computes;
 * an hour with no usable signal stays null and renders as an empty cell — never zero-filled, never
 * invented. Bands match the shared 0–3 scale: <1 low, <2 medium, >=2 high.
 *
 * APPROXIMATE and non-clinical, exactly like the underlying score: a wellness pattern view of an
 * autonomic-load proxy, never a stress diagnosis and never a medical reading.
 */
object StressHeatmap {

    /** Upper bound (exclusive) of the LOW band on the shared 0–3 scale. */
    const val LOW_BAND_CEILING: Double = 1.0
    /** Upper bound (exclusive) of MEDIUM; at/above this is HIGH (= DaytimeStress high band floor). */
    const val HIGH_BAND_FLOOR: Double = 2.0

    /** Default row window — the same waking band the intraday timeline uses (06:00–22:00). */
    const val DEFAULT_START_HOUR: Int = 6
    const val DEFAULT_END_HOUR: Int = 22

    enum class Band { LOW, MEDIUM, HIGH }

    /** Band for a 0–3 level. Matches the UI's StressBand cut points exactly. */
    fun band(level: Double): Band = when {
        level < LOW_BAND_CEILING -> Band.LOW
        level < HIGH_BAND_FLOOR -> Band.MEDIUM
        else -> Band.HIGH
    }

    /**
     * One day's hourly levels: [levels] maps local hour (0..23) -> 0–3 autonomic load. Hours absent from
     * the map are treated as no-data. Callers build this from the intraday timeline.
     */
    data class DayColumn(val day: String, val levels: Map<Int, Double>)

    /** One grid cell. [level] == null means no usable signal for that day/hour (render empty). */
    data class Cell(
        val day: String,
        val dayIndex: Int,
        val hour: Int,
        val level: Double?,
    ) {
        /** Band for the cell, or null when there's no level (never guesses a band). */
        val band: Band? get() = level?.let { band(it) }
    }

    /** Pattern read-out over a grid, with honest coverage so the UI can gate the narrative. */
    data class Summary(
        val peakHour: Int?,
        val peakHourMean: Double?,
        val calmestHour: Int?,
        val calmestHourMean: Double?,
        val overallMean: Double?,
        val coverage: Double,
        val scoredCells: Int,
        val totalCells: Int,
    ) {
        companion object {
            val EMPTY = Summary(null, null, null, null, null, 0.0, 0, 0)
        }
    }

    /**
     * Build the grid in row-major order (for each hour, each column) so a UI can lay out rows directly.
     * Hours outside [startHour, endHour) are omitted; an inverted/empty window yields an empty list.
     */
    fun grid(
        columns: List<DayColumn>,
        startHour: Int = DEFAULT_START_HOUR,
        endHour: Int = DEFAULT_END_HOUR,
    ): List<Cell> {
        val lo = startHour.coerceIn(0, 23)
        val hi = endHour.coerceIn(0, 24)
        if (hi <= lo || columns.isEmpty()) return emptyList()
        val out = ArrayList<Cell>((hi - lo) * columns.size)
        for (hour in lo until hi) {
            columns.forEachIndexed { i, col ->
                out.add(Cell(day = col.day, dayIndex = i, hour = hour, level = col.levels[hour]))
            }
        }
        return out
    }

    /**
     * Mean level per hour-of-day across all columns (only scored cells contribute). Hours with no scored
     * cell are ABSENT from the result — never zero-filled.
     */
    fun meanByHour(cells: List<Cell>): Map<Int, Double> {
        val sums = HashMap<Int, Double>()
        val counts = HashMap<Int, Int>()
        for (c in cells) {
            val l = c.level ?: continue
            sums[c.hour] = (sums[c.hour] ?: 0.0) + l
            counts[c.hour] = (counts[c.hour] ?: 0) + 1
        }
        val out = HashMap<Int, Double>()
        for ((hour, sum) in sums) {
            val n = counts[hour] ?: 0
            if (n > 0) out[hour] = sum / n
        }
        return out
    }

    /**
     * Pattern summary over a grid. Ties on peak/calmest resolve to the EARLIER hour so the result is
     * deterministic and identical to the Swift twin.
     */
    fun summary(cells: List<Cell>): Summary {
        if (cells.isEmpty()) return Summary.EMPTY
        val byHour = meanByHour(cells)
        val scored = cells.mapNotNull { it.level }
        val total = cells.size
        if (scored.isEmpty() || byHour.isEmpty()) {
            return Summary(null, null, null, null, null, 0.0, 0, total)
        }
        // Deterministic: ascending by (mean, hour) — first = calmest, and the peak takes the highest mean
        // with the EARLIEST hour among equals.
        val ascending = byHour.entries.sortedWith(
            compareBy({ it.value }, { it.key }),
        )
        val calmest = ascending.first()
        val maxMean = ascending.last().value
        val peak = ascending.filter { it.value == maxMean }.minByOrNull { it.key }!!
        val overall = scored.sum() / scored.size
        return Summary(
            peakHour = peak.key,
            peakHourMean = peak.value,
            calmestHour = calmest.key,
            calmestHourMean = calmest.value,
            overallMean = overall,
            coverage = scored.size.toDouble() / total.toDouble(),
            scoredCells = scored.size,
            totalCells = total,
        )
    }
}

File: Packages/StrandAnalytics/Sources/StrandAnalytics/StressHeatmap.swift
import Foundation

// StressHeatmap.swift — pure week×hour aggregation of the intraday autonomic-load timeline (R7).
//
// The Autonomic Load screen already shows TODAY's hour-by-hour read (`DaytimeStress`) and a daily trend.
// This engine answers the pattern question those two can't: **which hours of which days are habitually
// loaded?** ("Wednesday afternoons run hot", "you unwind after 21:00"). That is the coachable insight a
// heatmap exists for, and it is derived entirely from data NOOP already computes — no new sensor, no new
// claim, nothing fabricated: an hour with no usable signal stays `nil` and renders as an empty cell.
//
// Pure + closed-form so it is unit-tested and BYTE-IDENTICAL to the Android twin
// (com.noop.analytics.StressHeatmap). Bands match the app's shared 0–3 scale (StressBand): <1 low,
// <2 medium, ≥2 high.
//
// APPROXIMATE and non-clinical, exactly like the underlying score: this is a wellness pattern view of an
// autonomic-load proxy, never a stress diagnosis and never a medical reading.
public enum StressHeatmap {

    // MARK: - Bands (mirror StressBand / DaytimeStress.highBandFloor)

    /// Upper bound (exclusive) of the LOW band on the shared 0–3 scale.
    public static let lowBandCeiling: Double = 1.0
    /// Upper bound (exclusive) of the MEDIUM band; at/above this is HIGH (= DaytimeStress.highBandFloor).
    public static let highBandFloor: Double = 2.0

    public enum Band: String, Equatable, Sendable {
        case low, medium, high
    }

    /// Band for a 0–3 level. Matches the UI's `StressBand(score:)` cut points exactly.
    public static func band(_ level: Double) -> Band {
        if level < lowBandCeiling { return .low }
        if level < highBandFloor { return .medium }
        return .high
    }

    // MARK: - Input / output

    /// One day's hourly levels: `levels[h]` is the 0–3 autonomic load for local hour `h` (0…23), or nil
    /// when that hour had no usable signal. Callers build this from `DaytimeStress.Result.hours`.
    public struct DayColumn: Equatable, Sendable {
        /// Local day key, `yyyy-MM-dd`.
        public let day: String
        /// Sparse hour → level map (0…23 keys). Hours absent from the map are treated as no-data.
        public let levels: [Int: Double]
        public init(day: String, levels: [Int: Double]) {
            self.day = day
            self.levels = levels
        }
    }

    /// One grid cell. `level == nil` ⇒ no usable signal for that day/hour (render as an empty slot).
    public struct Cell: Equatable, Sendable {
        public let day: String
        /// Column index in the supplied order (0 = first/oldest column given).
        public let dayIndex: Int
        /// Local hour-of-day row, 0…23.
        public let hour: Int
        public let level: Double?
        public init(day: String, dayIndex: Int, hour: Int, level: Double?) {
            self.day = day
            self.dayIndex = dayIndex
            self.hour = hour
            self.level = level
        }
        /// Band for the cell, or nil when there's no level (never guesses a band).
        public var band: Band? { level.map { StressHeatmap.band($0) } }
    }

    /// The pattern read-out over a grid: which hour runs hottest/calmest across the window, plus honest
    /// coverage so the UI can gate the narrative until there is enough data.
    public struct Summary: Equatable, Sendable {
        /// Hour-of-day (0…23) with the highest mean level, or nil when nothing is scored.
        public let peakHour: Int?
        /// Mean level at `peakHour`.
        public let peakHourMean: Double?
        /// Hour-of-day with the lowest mean level, or nil.
        public let calmestHour: Int?
        /// Mean level at `calmestHour`.
        public let calmestHourMean: Double?
        /// Mean level across every scored cell, or nil.
        public let overallMean: Double?
        /// Scored cells ÷ total cells, 0…1 — the honesty gate (sparse wear ⇒ low coverage).
        public let coverage: Double
        public let scoredCells: Int
        public let totalCells: Int
        public init(peakHour: Int?, peakHourMean: Double?, calmestHour: Int?, calmestHourMean: Double?,
                    overallMean: Double?, coverage: Double, scoredCells: Int, totalCells: Int) {
            self.peakHour = peakHour
            self.peakHourMean = peakHourMean
            self.calmestHour = calmestHour
            self.calmestHourMean = calmestHourMean
            self.overallMean = overallMean
            self.coverage = coverage
            self.scoredCells = scoredCells
            self.totalCells = totalCells
        }
        /// Empty read — no columns or nothing scored.
        public static let empty = Summary(peakHour: nil, peakHourMean: nil, calmestHour: nil,
                                          calmestHourMean: nil, overallMean: nil, coverage: 0,
                                          scoredCells: 0, totalCells: 0)
    }

    // MARK: - Grid

    /// Default row window: the same waking band the intraday timeline uses (06:00–22:00 inclusive of 6,
    /// exclusive of 22), so the heatmap and the timeline agree on what a "day" covers.
    public static let defaultStartHour: Int = DaytimeStress.wakingStartHour
    public static let defaultEndHour: Int = DaytimeStress.wakingEndHour

    /// Build the grid in row-major order (for each hour, each column) so a UI can lay out rows directly.
    /// Hours outside `[startHour, endHour)` are omitted. An out-of-order or clamped window yields [].
    public static func grid(columns: [DayColumn],
                            startHour: Int = defaultStartHour,
                            endHour: Int = defaultEndHour) -> [Cell] {
        let lo = max(0, min(23, startHour))
        let hi = max(0, min(24, endHour))
        guard hi > lo, !columns.isEmpty else { return [] }
        var out: [Cell] = []
        out.reserveCapacity((hi - lo) * columns.count)
        for hour in lo..<hi {
            for (i, col) in columns.enumerated() {
                out.append(Cell(day: col.day, dayIndex: i, hour: hour, level: col.levels[hour]))
            }
        }
        return out
    }

    /// Mean level per hour-of-day across all columns (only scored cells contribute). Hours with no scored
    /// cell are absent from the result — never zero-filled.
    public static func meanByHour(_ cells: [Cell]) -> [Int: Double] {
        var sums: [Int: Double] = [:]
        var counts: [Int: Int] = [:]
        for c in cells {
            guard let l = c.level else { continue }
            sums[c.hour, default: 0] += l
            counts[c.hour, default: 0] += 1
        }
        var out: [Int: Double] = [:]
        for (hour, sum) in sums {
            if let n = counts[hour], n > 0 { out[hour] = sum / Double(n) }
        }
        return out
    }

    /// Pattern summary over a grid. Ties on peak/calmest resolve to the EARLIER hour so the result is
    /// deterministic (and identical to the Kotlin twin).
    public static func summary(_ cells: [Cell]) -> Summary {
        guard !cells.isEmpty else { return .empty }
        let byHour = meanByHour(cells)
        let scored = cells.compactMap(\.level)
        let total = cells.count
        guard !scored.isEmpty, !byHour.isEmpty else {
            return Summary(peakHour: nil, peakHourMean: nil, calmestHour: nil, calmestHourMean: nil,
                           overallMean: nil, coverage: 0, scoredCells: 0, totalCells: total)
        }
        // Deterministic tie-break: sort by (mean, hour) and take the ends.
        let ascending = byHour.sorted { a, b in
            a.value == b.value ? a.key < b.key : a.value < b.value
        }
        let calmest = ascending.first
        // For the peak, prefer the highest mean but the EARLIEST hour among equals.
        let peak = ascending.reversed().min { a, b in
            a.value == b.value ? a.key < b.key : a.value > b.value
        }
        let overall = scored.reduce(0, +) / Double(scored.count)
        return Summary(peakHour: peak?.key,
                       peakHourMean: peak?.value,
                       calmestHour: calmest?.key,
                       calmestHourMean: calmest?.value,
                       overallMean: overall,
                       coverage: Double(scored.count) / Double(total),
                       scoredCells: scored.count,
                       totalCells: total)
    }
}
